makemake - BlockFeatured & GridFeatured

austinblanchard opened the pull request:
Components included in this PR are:

1. BlockFeatured - With logos / credits (+hover state) / news variants
1. GridFeatured

drewbaker commented on src/components/BlockFeatured/BlockFeatured.vue:
```diff
@@ -122,6 +160,42 @@ export default {
         },
         formattedCredits() {
             return nl2br(this.credits).split("<br>")
+        },
+        formattedCreditsSecondary() {
+            return nl2br(this.creditsSecondary).split("<br>")
+        }
```
You don't need `nl2br` here if you are doing the split. You can just do `this.credits.split(/(?:\r\n|\r|\n)/g)`

drewbaker commented on src/components/BlockFeatured/BlockFeatured.vue:
```diff
@@ -100,7 +128,11 @@ export default {
             default: ""
         },
-        logos: {
+        creditsSecondary: {
```
Lets have `creditsPrimary` and `creditsSecondary` as props.

drewbaker commented on src/components/BlockFeatured/BlockFeatured.vue:
```diff
@@ -122,6 +160,42 @@ export default {
+    },
+    methods: {
+        startHover() {
+            if (this.credits) {
+                // Move image up the height of the credits block
+                this.$el.querySelector(
+                    ".media"
+                ).style.transform = `translateY(-${this.imageTranslateValue})`
+                // Animate height of credits spacer to match visible credits
+                this.$el.querySelector(
+                    ".additional-credits"
+                ).style.maxHeight = `${this.creditsHeight}`
+                // Animate background hover state credits up from translated position
+                this.$refs.secondary.style.transform = "none"
```
Don't use JS to set the style like this. You want to do this using data and a computed style (kind of like we do for for classes).

So you have the same methods, but the methods sets data() of the height, then you have a computed style, like this: https://github.com/funkhaus/liveinithaca.com/blob/617a38a71d451bc0e99dbb8db9f59cb06f434e39/src/components/block/ExpandedInfo.vue#L45-L51

That way your method just changes a data() val, and everything from there is Vue reactive.

If you do the z-index approach on the credits, then you also won't need the transform: none.

austinblanchard replied:
@drewbaker Ok, I've switched the ones that have accessible markup in the component, so I can use `:style` in the way you mentioned. But for translating the image, I'm doing that on the `img` element within WpImage, which I don't have access to in the component, and therefore was targeting it using querySelector(). Do you prefer another way of doing this, or does that work?

drewbaker replied:
@austinblanchard Why don't you just translate the entire `wp-image`? Why does it need to be `.media` inside it? 

```
        <wp-image
            v-if="image && !isNews"
            class="block-image"
            :image="image"
            mode="fullbleed"
           :style="imageStyles"
        />

       // Computed
       imageStyles() {
            return {
                 transform: translateY(this.imageTranslateValue);
            }
      }
```

drewbaker commented on src/components/BlockFeatured/BlockFeatured.vue:
```diff
@@ -133,10 +207,11 @@ export default {
     .block-image {
         .media {
-            transition: top 0.5s $authenticMotion;
+            transition: transform 0.5s $authenticMotion;
```
Lets use 0.4s on all the times for this.

drewbaker commented on src/components/BlockFeatured/BlockFeatured.vue:
```diff
@@ -159,65 +234,87 @@ export default {
+            opacity: 1;
         }
     }
+    .logo-jax {
+        height: 17px;
+    }
+    .logo-rpsg {
+        height: 17px;
+    }
+    .logo-a52 {
+        height: 15px;
+    }
+    .logo-a52-color {
+        height: 17px;
+    }
+    .logo-elastic {
+        height: 12px;
+    }
+    .logo-indestrucible {
+        height: 10px;
+    }
```
Just have the height be 17px on all logos, width: auto.

drewbaker commented on src/components/BlockFeatured/BlockFeatured.vue:
```diff
@@ -1,55 +1,71 @@
             :image="image"
             mode="fullbleed"
         >
         </wp-image>
```
You don't need this last line. Just self close  `<wp-image/>`

drewbaker commented on src/styles/global.scss:
```diff
@@ -11,6 +11,7 @@
     --unit-100vh: 100vh;
     --unit-vertical-gap: 70px;
     --unit-gutter: 50px;
+    --unit-spacer: 17px;
```
Don't think we need this as a global, this can be defined only in your component if you need it.

drewbaker commented on src/styles/global.scss:
```diff
@@ -25,3 +26,11 @@ a {
 }
+
+// SVG styles
+svg {
+    path,
+    circle {
+        fill: var(--color-black);
+    }
+}
```
Make this `.svg` so that it's opt in for all SVGs with that class.

drewbaker commented on src/components/GridFeatured/GridFeatured.stories.js:
```diff
@@ -0,0 +1,93 @@
+        credits: `Artists – Angus Wall & Hazel Baird
+        Editor – Jamie Foord`,
+        creditsSecondary: `Production Co. – Placeholder
+        Agency – Placeholder`
+    })
+    .concat({
+        ...API.page,
+        isNews: true,
+        title:
+            "Mikkel EG Nielsen’s first spot for the new year is an inverted twist to the story of Narcissus"
+    })
+    .concat({
+        ...API.pages[2],
+        hasHover: true,
+        credits: `Artists – Angus Wall & Hazel Baird
+        Agency – Placeholder`,
+        creditsSecondary: `Production Co. – Placeholder
+        Agency – Placeholder`
+    })
+    .concat({
+        ...API.page,
+        hasHover: true,
+        credits: `Artists – Angus Wall & Hazel Baird
+        Production Co. – Placeholder`,
+        creditsSecondary: `Editor – Jamie Foord
+        Agency – Placeholder`
+    })
+    .concat({
+        ...API.page,
+        isNews: true,
+        title:
+            "A whole lot more than deep dish pizza in Nike’s ‘Chicago Style.’"
+    })
+    .concat({
+        ...API.pages[0],
+        hasHover: true,
+        credits: `Artists – Angus Wall & Hazel Baird
+        Editor – Jamie Foord
+        Agency – Placeholder`
+    })
```
FYI a more efficient way to do this would be to do this. Not super important but just thought I'd mention it.

```
let item = {
        ...API.pages[0],
        hasHover: true,
        credits: `Artists – Angus Wall & Hazel Baird
        Editor – Jamie Foord`,
        creditsSecondary: `Production Co. – Placeholder
        Agency – Placeholder`
}
const newsAndCredits = [
     {...item},
     {
          ...item,
         image: API.pages[1].featuredImage
        creditsSecondary: `Editor – Jamie Foord
        Agency – Placeholder`          
     }
]

```

drewbaker commented on src/components/GridFeatured/GridFeatured.vue:
```diff
@@ -0,0 +1,57 @@
+            :key="i"
+            :to="block.uri"
+            :is-news="block.isNews"
+            :has-hover="block.hasHover"
```
This will actually be a method, based on the `postType` from the API I think, or maybe the GQL `__typename`, because sometimes they link to a page or a post. So can you have this call a method `hasHover(block)` and that just returns `return block.hasHover`for now and I'll update once the API is built.

drewbaker commented on src/components/GridFeatured/GridFeatured.vue:
```diff
@@ -0,0 +1,57 @@
+    <div class="grid-featured">
+        <block-featured
+            v-for="(block, i) in blocks"
+            :key="i"
```
Don't use index as key! Use `block.id`. Yes this will through errors in Storybook, that is OK.

![1bb1952285f30288900a434ddce22e52](https://user-images.githubusercontent.com/475015/86998763-ea996300-c165-11ea-88c7-5dd1b1c03394.jpg)

austinblanchard commented:
"Then on hover you just bump up the image, **translate meta**, and fade in play icon."

@drewbaker Responding here to some of your notes above because there wasn't a way for me to comment up there. Regarding the note to translate meta - I tried this initially, but learned (when it didn't work) that setting `transform` to anything other than `none` results in creating a new [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context), therefore nullifying our z-index setup for the hover state, so I used negative margin-bottom instead (not ideal, but it works). Let me know if that seems alright.

...

"See all my review notes in GitHub too. I don't think you need to give each SVG a class name in the template. If you need a class name on it, then you can do that on the `<svg-logo-name class="some-class"/>` you don't need it in the SVG file."

@drewbaker  These logos are rendered in a loop, so there's no way to add the class in the markup. That's why I put them in the SVG files, because it's common to need to tweak some so they look the right size together. I made the default 17px, so only the outliers require additional styling. Let me know if you think I still shouldn't have the classes in the SVG files.

drewbaker commented:
Can you also change the top level `nuxt-link` to a `smart-link` (add it to global) as they might link these out of the site entirely. 
https://github.com/funkhaus/factory/blob/40812901c4e1cd3ef43dca7d214c558d7b4bb8a8/src/components/BlockFeatured/BlockFeatured.vue#L2

austinblanchard commented:
@drewbaker Made the rest of the changes discussed here.

drewbaker commented:
If I understand that stacking content correctly, can you put a translate of 0 on the parent, and then use translates in child components with the right z-indexs? Otherwise can you use `position:relative` and top or bottom instead of negative margin? It's just a little more intuitive than a negative margin.

You can totally apply the class in a loop.

```
                    <component
                        :is="`svg-logo-${name}`"
                        v-for="(name, i) in logoNames"
                        :key="name"
                        :class="`logo logo-${name}`"
                    />
```

Please do not hard code classes in the SVGs ever. The only time to do something like thta was if you wanted to target specific things inside an SVG (like one path was blue, and another was red).

austinblanchard commented:
@drewbaker Of course, you're right about the SVG classes in the loop, my bad. I've removed them from the SVG files. Regarding the stacking context, I didn't quite understand your takeaway with putting translate of 0 on the parent, but I tried it and it didn't seem to fix the issue. I've updated the code to use `top` rather than `margin-bottom` as you asked.

drewbaker commented:
I'm working on this now, it's 100% possible to do this with only translates.
